part3/MSI/debugging.rst

```diff
@@ -61,6 +62,59 @@ Once you have your protocol working with the random tester, you can move on to u
 It is likely that real applications will expose even more bugs in the protocol.
 If at all possible, it is much easier to debug your protocol with the random tester than with real applications!
 
+Understanding Protocol Traces
+=============================
+
+Unfortunately, despite extensive effort to catch bugs in them, coherence protocols (even heavily tested ones) will have sometimes have bugs.
+Sometimes these bugs are relatively simple fixes, while other times the bugs will be very insidious and difficult to track down.
+In the worst case, the bugs will manifest themselves as deadlocks: bugs that literally prevent the application from making progress.
+Another, similar problem is livelocks: where the program runs forever because there is a cycle somewhere in the system.
+Whenever livelocks or deadlocks occur, the next thing to do is generate a protocol trace.
+Traces print a running list of every transition that is happening in the memory system -- memory requests starting and completing, L1 and directory transitions, etc.

```

Comment:
Can you insert a "real" em-dash here (—)? Apparently that' what restructuredtext expects.